packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +959,1336 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+  }
+}
+
+class _RenderTheatreMarkerElement extends InheritedElement {
+  _RenderTheatreMarkerElement(super.widget);
+
+  _RenderTheatreMarkerElement get _topRenderTheatre {
+    _RenderTheatreMarkerElement? ancestor;
+    visitAncestorElements((Element element) {
+      assert(ancestor == null);
+      ancestor = element.getElementForInheritedWidgetOfExactType<_RenderTheatreMarker>() as _RenderTheatreMarkerElement?;
+      return false;
+    });
+    return ancestor?._topRenderTheatre ?? this;
+  }
+}
+
+class _OverlayPortalState extends State<OverlayPortal> {
+  // The developer must call `show` to reveal the overlay so we can get a unique
+  // timestamp of the user interaction for sorting.
+  _OverlayEntryLocation? _location;
+  int? _showTimestamp;
+
+  @override
+  void initState() {
+    super.initState();
+    _setupController(widget.controller);
+  }
+
+  void _setupController(OverlayPortalController controller) {
+    assert(
+      controller._attachTarget == null || controller._attachTarget == this,
+      'Failed to attach $controller to $this. It is already attached to ${controller._attachTarget}'
+    );
+    final int? controllerTimeStamp = controller._showTimestamp;
+    final int? timeStamp = _showTimestamp;
+    if (timeStamp == null || (controllerTimeStamp != null && controllerTimeStamp > timeStamp)) {
+      _showTimestamp = controllerTimeStamp;
+    }
+    controller._showTimestamp = null;
+    controller._attachTarget = this;
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _location?._dispose();
+    _location = null;
+  }
+
+  @override
+  void didUpdateWidget(OverlayPortal oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.controller != widget.controller) {
+      oldWidget.controller._attachTarget = null;
+      _setupController(widget.controller);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.controller._attachTarget = null;
+    super.dispose();
+  }
+
+  void show(int timestampe) {
+    assert(SchedulerBinding.instance.schedulerPhase != SchedulerPhase.persistentCallbacks);
+    setState(() { _showTimestamp = timestampe; });
+    _location?._dispose();
+    _location = null;
+  }
+
+  void hide() {
+    assert(SchedulerBinding.instance.schedulerPhase != SchedulerPhase.persistentCallbacks);
+    setState(() { _showTimestamp = null; });
+    _location = null;
+    _location?._dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final int? timestamp = _showTimestamp;
+    if (timestamp == null) {
+      return _OverlayPortal(
+        overlayLocation: null,
+        overlayChild: null,
+        child: widget.child,
+      );
+    }
+    return _OverlayPortal(
+      overlayLocation: _location ??= _OverlayEntryLocation(timestamp, widget._resolveLocation(context)._overlayEntryTopChildModel),
+      overlayChild: _DeferredLayout(child: widget.overlayChildBuilder(context)),
```

Comment:
Should overlayChildBuilder be wrapped in a Builder instead of passing in the context of the _OverlayPortalState to allow for more targeted rebuilds if necessary?

Reply:
Should I do that for `OverlayEntry.builder` too?